go/apps/mixnet/proxy.go

```diff
@@ -72,51 +83,99 @@ func (p *ProxyContext) SendMessage(c net.Conn, msg []byte) (int, error) {
 
 	bytes := copy(cell[1+n:], msg)
 	if _, err := c.Write(cell); err != nil {
-		return 0, err
+		return err
 	}
 
 	for bytes < msgBytes {
 		zeroCell(cell)
-		bytes += copy(cell, msg[bytes:])
+		cell[0] = msgCell
+		bytes += copy(cell[1:], msg[bytes:])
 		if _, err := c.Write(cell); err != nil {
-			return bytes, err
+			return err
 		}
 	}
-
-	return bytes, nil
+	return nil
 }
 
-// ReceiveMessage waits for a reply or error message from the router.
-func (p *ProxyContext) ReceiveMessage(c net.Conn, msg []byte) (int, error) {
+// ReceiveMessage reads message cells from the router and assembles them into
+// a messsage.
+func (p *ProxyContext) ReceiveMessage(c *Conn) ([]byte, error) {
 	var err error
+
+	// Receive cells from router.
 	cell := make([]byte, CellBytes)
 	if _, err = c.Read(cell); err != nil && err != io.EOF {
-		return 0, err
+		return nil, err
 	}
 
-	if cell[0] == msgCell { // Read a message.
-		// TODO(cjpatton)
-
-	} else if cell[0] == dirCell { // Handle a directive.
-		dirBytes, n := binary.Uvarint(cell[1:])
+	if cell[0] == dirCell {
 		var d Directive
-		if err := proto.Unmarshal(cell[1+n:1+n+int(dirBytes)], &d); err != nil {
-			return 0, err
+		err = unmarshalDirective(cell, &d)
```

Comment:
style: fold this statement into the following if statement, and break out the else-if into its own if statement